state-commitments/optimized/history_commitment.go

```diff
@@ -50,18 +77,43 @@ func NewCommitter() *HistoryCommitter {
 	}
 }
 
+func (h *HistoryCommitter) hash(item []byte) (common.Hash, error) {
+	defer h.keccak.Reset()
+	if _, err := h.keccak.Write(item); err != nil {
+		return emptyHash, err
+	}
+	var result common.Hash
+	if _, err := h.keccak.Read(result[:]); err != nil {
+		return emptyHash, err
+	}
+	return result, nil
+}
+
 func (h *HistoryCommitter) ComputeRoot(leaves []common.Hash, virtual uint64) (common.Hash, error) {
 	if len(leaves) == 0 {
-		return common.Hash{}, nil
+		return emptyHash, nil
 	}
-	// Called with 0 limit first to compute the last leaf fillers for the commitment.
-	copiedLeaves := make([]common.Hash, len(leaves))
-	copy(copiedLeaves, leaves)
-	return h.computeVirtualSparseTree(copiedLeaves, virtual, 0)
+	rehashedLeaves := make([]common.Hash, len(leaves))
+	for i, leaf := range leaves {
+		result, err := h.hash(leaf[:])
+		if err != nil {
+			return emptyHash, err
+		}
+		rehashedLeaves[i] = result
+	}
+	return h.computeVirtualSparseTree(rehashedLeaves, virtual, 0)
 }
 
 func (h *HistoryCommitter) GeneratePrefixProof(prefixIndex uint64, leaves []common.Hash, virtual uint64) ([]common.Hash, []common.Hash, error) {
-	prefixExpansion, proof, err := h.prefixAndProof(prefixIndex, leaves, virtual)
+	rehashedLeaves := make([]common.Hash, len(leaves))
+	for i, leaf := range leaves {
+		result, err := h.hash(leaf[:])
+		if err != nil {
+			return nil, nil, err
+		}
+		rehashedLeaves[i] = result
```

Comment:
This might benefit from a hashLeaves function to reduce code reuse

Reply:
Done.